### DISCUSSION TITLE
offset() within Locations context

### DISCUSSION THREAD

**Discussion by nanosonde** [gumyr/build123d#434] [Q&A] [answered]
Hi!
I am new to build123d and the CAD world.

I am currently trying to do the  following:
```
with BuildPart() as mypart:
    Box(width, height, thickness)

    with Locations(mypart.faces().sort_by(Axis.Z).last):
        with Locations((0, 10, (depth/2))):
            Box(width, window_height+thickness, depth)
            offset(amount=-thickness, openings=mypart.faces().sort_by(Axis.Z)[-1])
```

The first box is supposed to be some base plate where I would like to place a shell on top.

When I try to compile this, I get the following error:
```
    offset(amount=-thickness, openings=mypart.faces().sort_by(Axis.Z)[-1])
  File "C:\Users\User\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\build123d\operations_generic.py", line 633, in offset
    solid.offset_3d(openings_in_this_solid, amount, kind=kind).fix()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\User\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\build123d\topology.py", line 1284, in offset_3d   
    offset_occt_solid = offset_builder.Shape()
                       ^^^^^^^^^^^^^^^^^^^^^^
OCP.StdFail.StdFail_NotDone: BRep_API: command not done
```

When I removed the `Locations` context, I do not get an error. However, the position is of course not as intended.

So the questions is: is the `offset()`operation supported under a `Locations` context?

**Comment by gumyr** [comment]
`offset` is supported as used here. The problem here is likely that the offset can't be computed correctly given the geometry of your part.  It would help if you provided values for: `width`, `height`, `thickness`, `depth` and `window_height` so I could replicate exactly what you're trying to create.

**Reply by nanosonde** [reply]
Ok, thanks for replying back.

```
height = 120.0
width = 135.0
depth= 40.0
thickness = 5.0
window_width=125
window_height=95
```

**Comment by gumyr** [comment, marked as answer]
Ok, now I can see the problem:
![image](https://github.com/gumyr/build123d/assets/85638518/2188ed88-8660-4c4b-ac78-a0048776e1a7)

By specifying the offset amount as `-thickness` when the tab on the main box is `thickness` high the offset if impossible to create as the part's minimum dimension must be at least `2*thickness`.  If the offset amount is set to `-1`, it works as follows:
![image](https://github.com/gumyr/build123d/assets/85638518/101d2013-387a-4f76-bd33-f6ca97185349)
Notice how the offset is creating a hollow section in the small tab as well.

Good luck with your project.
